File: cpp/78_coin_partitions.cpp
#include <iostream>

int coin_partitions(){
    int partitions[60001] = {0};
    // one way to partition 0 coins
    partitions[0] = 1;
    for (int i = 1; i < 60001; i++){
        for (int j = i; j < 60001; j++){
            // create new partition of j by setting aside one pile of i coins
            // add partitions[j - i] ways to partition remaining j - i coins
            partitions[j] += partitions[j - i] % 1'000'000;
            partitions[j] %= 1'000'000;
        }
        // finished counting partitions for i coins
        // after setting aside 1, 2, 3, ..., i coins
        if (partitions[i] == 0) return i;
    }
    return -1;
}

int main(){
    std::cout << coin_partitions() << "\n";
}

File: c#/_78_CoinPartitions.cs
using System;

public class _78_CoinPartitions{
	public static int Euler78(){
		int[] partitions = new int[60_001];
		partitions[0] = 1;
		for (int i = 1; i < partitions.Length; i++){
			for (int j = i; j < partitions.Length; j++){
				partitions[j] += partitions[j - i] % 1_000_000;
				partitions[j] %= 1_000_000;
			}
			if (partitions[i] == 0) return i;
		}
		return 0;
	}
	
	public static void Main(){
		Console.WriteLine(Euler78());
	}
}
